File: cod_python/grafos.py
from collections import deque

def pessoa_e_vendedora(nome):
	return nome[-1] == 'm'

grafo = {}
grafo["voce"] = ["alice", "bob", "claire"]
grafo["bob"] = ["anuj", "peggy"]
grafo["alice"] = ["peggy"]
grafo["claire"] = ["thom", "jonny"]
grafo["anuj"] = []
grafo["peggy"] = []
grafo["thom"] = []
grafo["jonny"] = []


def pesquisa(nome):
	pesquisa_nome = deque()
	pesquisa_nome += grafo[nome]

	procurado = set()

	while pesquisa_nome:
		pessoa = pesquisa_nome.popleft()

		if pessoa not in procurado:
			if pessoa_e_vendedora(pessoa):
				print(pessoa + " e um vendedor de manga!")
				return True
			else:
				pesquisa_nome += grafo[pessoa]

				procurado.add(pessoa)

	return False

pesquisa("voce")

File: cod_javascript/grafos.js
const grafo = {}

grafo.voce = ['alice', 'bob', 'claire']
grafo.bob = ['anuj', 'peggy']
grafo.alice = ['peggy']
grafo.claire = ['thom', 'jonny']
grafo.anuj = []
grafo.peggy = []
grafo.thom = []

const eVendedor = name => name[name.length -1] === "m"

const pesquisa = (name, grafo = {}) => {
	const iterador = ( procurado = [], visitado) => {
		if(procurado.length === 0){
			return false
		}
		const [corrente, ...rest] = procurado

		if(visitado.has(corrente)){
			return iterador(rest, visitado)
		}

		if(eVendedor(corrente)){
			console.log(`${corrente} é vendedor de manga!`)
			return true
		}

		visitado.add(corrente)

		const pessoaAmiga = grafo[corrente]
	 	return iterador([...rest, ...pessoaAmiga], visitado)
	}
	return iterador(grafo[name], new Set())
}

console.log(pesquisa('voce', grafo))
